File: Server/src/main/java/org/spider/railway/utils/PID.java
package org.spider.railway.utils;

import lombok.extern.slf4j.Slf4j;

@Slf4j
public class PID {
    private final double Kp, Kd, Ki, min, max;
    private double prev, integral;
    private double target;
    private long prevMillis;

    public PID(double kp, double kd, double ki, double min, double max) {
        this.Kp = kp;
        this.Kd = kd;
        this.Ki = ki;
        this.min = min;
        this.max = max;

        this.prev = 0.0;
        this.prevMillis = 0;
        this.integral = 0.0;
        this.target = 0.0;
    }

    private double constraint(double value) {
        return Math.max(min, Math.min(value, max));
    }

    public synchronized void setTarget(double target) {
        this.target = target;
    }

    public synchronized double getResult(double current) {
        final double error = target - current;
        final long currentMillis = System.currentTimeMillis();

        //double prop = error * this.Kp, diff = 0.0, integr = 0.0;
        double result = error * this.Kp; // proportional

        if (prevMillis > 0) {
            final double delta = this.prev - current;
            final double deltaSec = (currentMillis - prevMillis) / 1000.0;
            //diff = delta * this.Kd / deltaSec;
            result += delta * this.Kd / deltaSec; // differential

            //integr = constraint(integral + error * Ki * deltaSec);
            this.integral = constraint(integral + error * Ki * deltaSec);
            result += integral;
        }

        //log.info("Prop={}, Diff={}, Integr={}.", prop, diff, integr);

        this.prevMillis = currentMillis;
        this.prev = current;

        return constraint(result);
    }
}

File: Embeded/pid.cpp
#include "esp_attr.h"
#include "esp_compiler.h"
#include "utils/Utils.h"
#include <pid\Pid.h>
#include <logger\logger.h>

static const char TAG[] PROGMEM = "PID";

float IRAM_ATTR PID::calc(const float current, const uint32_t currentMs) noexcept {
    const float error = target - current;

    float result = error * Kp; // proportional

    if (likely(prevMillis > 0)) {
        const float deltaSec = (currentMs - prevMillis) / 1000.0f;

        if (likely(deltaSec > 0.0f)) {
            const float delta = prev - current;

            result += delta * Kd / deltaSec; // differential

            integral = constraint(integral + error * Ki * deltaSec); // integral
            result += integral;
        }
    }

    prevMillis = currentMs;
    prev = current;

    return constraint(result);
}

PIDTask::PIDTask(PID& pid, PidSourceCallbackFunc sourceFunc, void *sourceArgs, PidDestinationCallbackFunc destinationFunc, void *destinationArgs,
                                    uint32_t pidStepMs, float minDiff) noexcept
    : QueueWorker<float>(1, "PID", NORMAL_TASK_PRIORITY),
        pid(pid),
        sourceFunc(sourceFunc),
        sourceArgs(sourceArgs),
        destinationFunc(destinationFunc),
        destinationArgs(destinationArgs),
        pidStepMs(pidStepMs), minDiff(minDiff), target(0)  {}

void IRAM_ATTR PIDTask::processQueueItem(float& value) noexcept {
    if (std::abs(value - target) >= minDiff) {
        do {
          rt:
            const bool isZero = std::abs(value) < minDiff;
            const float signValue = signNonZero(value);

            if (isZero || signValue != signNonZero(target)) { // stop
                pid.clear();
                const float safeTarget = target;
                setDestination(target = 0.0f);

                const uint32_t now = currentMillis();
                while (std::abs(getSource()) > minDiff) { // wait stop
                    if (unlikely(receiveQueueItem(value, pidStepMs) && std::abs(value - safeTarget) >= minDiff))
                        goto rt;

                    if (unlikely(currentMillis() - now > 10000)) {
                        SP_LOGW(TAG, "Timeout while wait stop.");
                        return;
                    }
                }

                if (unlikely(isZero))
                    return;
            }

            pid.setTarget(std::abs(target = value));
            do {
                setDestination(signValue * pid.calc(std::abs(getSource())));

                if (unlikely(receiveQueueItem(value, pidStepMs) && std::abs(value - target) >= minDiff))
                    break;
            } while (true);
        } while (true);
    }
}
